README.md
# weeclick-parse-server using Node.JS

## [Lien vers le Parse-dashboard actuel](https://weeclik-dash.herokuapp.com/)

Weeclik côté serveur utilise Node.JS avec le module : [parse-server](https://github.com/parse-community/parse-server) sur Express.

Il est actuellement hebergé sur des serveurs [Heroku](https://www.heroku.com/), qui est une solution d'hebergement Docker avec des addons gratuit axés sur la lourdeur des services.

La route /parse permet de fait les appels CRUD (Create, Read, Update, Delete)    
La route /dashboard permet d'afficher le dashboard 

La base de donnée est NoSQL avec [MongoDB](https://www.mongodb.com/), hebergé par [mLab](https://www.mlab.com/).    
Un dump de la base de donnée se trouve dans le dossier BD à la racine.  

Utiliser cette commande pour importer le dossier comportant les données de la bdd :
```
mongorestore -h <url-site-mlab>:<port> -d <nom_bdd> -u <utilisateur_admin_bdd> -p <mot_de_passe> <chemin_dossier/dump_bdd>
```

La documentation de parse server : https://docs.parseplatform.org/parse-server/guide/

Le module [parse-dashboard](https://github.com/parse-community/parse-dashboard) permet lui de visualiser et modifier les données de la BDD directement depuis un client web.

La mise en place du dashboard : https://github.com/parse-community/parse-dashboard   

### For Local Development

* Make sure you have at least Node 4.3. `node --version`
* Clone this repo and change directory to it.
* `npm install`
* Install mongo locally using http://docs.mongodb.org/master/tutorial/install-mongodb-on-os-x/
* Run `mongo` to connect to your database, just to make sure it's working. Once you see a mongo prompt, exit with Control-D
* Run the server with: `npm start`
* By default it will use a path of /parse for the API routes.  To change this, or use older client SDKs, run `export PARSE_MOUNT=/1` before launching the server.
* You now have a database named "dev" that contains your Parse data
* Install ngrok and you can test with devices

### Getting Started With Heroku + mLab Development

#### With the Heroku Button

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

#### Without It

* Clone the repo and change directory to it
* Log in with the [Heroku Toolbelt](https://toolbelt.heroku.com/) and create an app: `heroku create`
* Use the [mLab addon](https://elements.heroku.com/addons/mongolab): `heroku addons:create mongolab:sandbox --app YourAppName`
* By default it will use a path of /parse for the API routes.  To change this, or use older client SDKs, run `heroku config:set PARSE_MOUNT=/1`
* Deploy it with: `git push heroku master`

# Using it

You can use the REST API, the JavaScript SDK, and any of our open-source SDKs:

Example request to a server running locally:

```curl
curl -X POST \
  -H "X-Parse-Application-Id: myAppId" \
  -H "Content-Type: application/json" \
  -d '{"score":1337,"playerName":"Sean Plott","cheatMode":false}' \
  http://localhost:1337/parse/classes/GameScore
  
curl -X POST \
  -H "X-Parse-Application-Id: myAppId" \
  -H "Content-Type: application/json" \
  -d '{}' \
  http://localhost:1337/parse/functions/hello
```

Example using it via JavaScript:

```javascript
Parse.initialize('myAppId','unused');
Parse.serverURL = 'https://whatever.herokuapp.com';

var obj = new Parse.Object('GameScore');
obj.set('score',1337);
obj.save().then(function(obj) {
  console.log(obj.toJSON());
  var query = new Parse.Query('GameScore');
  query.get(obj.id).then(function(objAgain) {
    console.log(objAgain.toJSON());
  }, function(err) {console.log(err); });
}, function(err) { console.log(err); });
```

Example using it on Android:
```java
//in your application class

Parse.initialize(new Parse.Configuration.Builder(getApplicationContext())
  .applicationId("myAppId")
  .server("http://myServerUrl/parse/")   // '/' important after 'parse'
  .build());

ParseObject testObject = new ParseObject("TestObject");
testObject.put("foo", "bar");
testObject.saveInBackground();
```
Example using it on iOS (Swift):
```swift
//in your AppDelegate

Parse.initializeWithConfiguration(ParseClientConfiguration(block: { (configuration: ParseMutableClientConfiguration) -> Void in
  configuration.server = "https://<# Your Server URL #>/parse/" // '/' important after 'parse'
  configuration.applicationId = "<# Your APP_ID #>"
}))
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.        
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]
### Changed
- When charging a commerce via Stripe API, commerceId is now being passed as a parameter
- When sharing a commerce, a timer is also set on Back to prevent looping on endpoint and getting an infinite number of sharing

## [2.3.2] - 2020-12-25
### Changed
- Bugfixes for endpoint POST (/publish-commerce).
- Added required user id for endpoint POST (/publish-commerce).

## [2.3.1] - 2020-12-25
### Changed
- Fix async keyword missing in /publish-commerce endpoint

## [2.3] - 2020-12-25
### Added
- Added a new endpoint POST (/publish-commerce), to publish a commerce to weeclik network via backend, much secure than writing business logic in react app

### Changed
- Update cloud function after save for commerce, now create an empty array for stripeCheckoutSession, storing stripe checkout sessions IDs to prevent fraud.

## [2.2] - 2020-11-19
### Added
- Added a new endpoint POST (/create-checkout-session), for payment checkout / CSA 3D secure integration
- Added a new endpoint POST (/retrieve-checkout-session-status), to confirm the successfull payment of a checkout sssion

### Changed
- Move someroute to different files, all routes / cron are no longer in one unique file

### Removed
- Endpoint POST (/charge) since no longer used by web app project

## [2.1] - 2020-11-08
### Changed
- Prepare MongoDB Atlas migration with removal of MONGODB_URI config var, the server app now uses DATABASE_URI config var name (not linked to Heroku's MongoDB Add-on).

## [2.0] - 2020-09-24
### Added
- New sharing Endpoint POST (/share). Needs a commerceId(required) and a userId(optional) to increment sharing number.
- After save of commerce now check if end of subscription is null, set its date tu current, so the cron task, checking for expired commerces), can invalidate it directly.

### Changed
- After save of commerce now check if it's null not only if latitude & longitude are egal to 0.
- Better handling of [Stripe](https://stripe.com/fr) payment methods (billing + others) - POST: /charge (async).
- Clean up of dependencies
- Updated npm librairies such as Parse, Dashboard, etc.

### Removed
- After save of commerce no longer check description for hashtags.
- Endpoint GET (/valid-email/:email), for checking if user email is valid, never used.
- Endpoint GET (/redirect-to-store), for getting a link to download the app, [Weeclik website](https://www.weeclik.com/) is now being used for this purpose.
- Endpoint POST (/webhook), which was used during [Stripe](https://stripe.com/fr) integration tests.

## [1.0.7] - 2020-01-12
### Added
- Add [Stripe](https://stripe.com/fr) charge method (unique payment). - POST: /charge (async).
- Cloud code function (endedSubscription) to track ended payments of commerces hourly.
- Added endpoint for universal links (iOS). - GET /apple-app-site-association.
- Added Parse Cloud AfterDelete function to remove all commerce's photos & videos if deleted.
- Prevent .DS_Store files to be tracked by Git.

### Changed
- Major update of parse-server dependency from v2 to v3.
- Parse server cron (automatic schedule) now execute new cloud code function to retrieve commerce's ended subscriptions.
- Better Parse Cloud AfterSave function.
- Added other banned word to commerce's tags ("description", "sappuie", "sur", "pour", "les", "proposer", "très").

### Removed
- Cloud code function (retrieveAllObjects) since parse-server framework changed from v2 to v3 (function no longer can be executed).

## [1.0.6] - 2019-12-01
### Changed
- Change handling of /charge endpoint.

## [1.0.5] - 2019-11-24
### Added
- Add CORS domain origin option on POST (/charge).

## [1.0.4] - 2019-11-16
### Added
- Created CHANGELOG.md file to keep track of changes.

## [1.0.3] - 2019-11-16
### Added
- Increase max file upload size to 1Go. 

## [1.0.2] - 2019-07-14
### Removed
- Removed Javascript key from Parse Server configuration (No longer needed).

cron/cron.js
module.exports = function(app){

  const cron          = require('node-cron');
  let moment          = require('moment');
  let Parse           = require('parse/node');

  // Toute les heure à 0 minute
  // '* * 1 * *' tous les mois ?
  // '* * * * *'
  // cron.schedule('*/10 * * * * *', async () => {
  cron.schedule('0 * * * *', async () => {
  // Chaque heure à 0 (01:00, 15:00, etc)
    // TODO: Ecrire fonction pour ecriture de log
    // TODO: Envoi de mail à chaque commerce passant en mode desactivé (user & admin)
    const functionName = 'endedSubscription';
    console.log(`Function ${functionName} executé à ${moment()}`);
    const commerces = await Parse.Cloud.run(functionName);
    console.log("Successfully retrieved " + commerces.length + " commerces.");

    for (let i = 0; i < commerces.length; i++) {
      let object = commerces[i];
      if (moment(object.get('endSubscription')).isValid()) {
        let day =  moment(object.get('endSubscription'));
        if (moment().isSameOrAfter(day)) {
          console.log(object.get('nomCommerce') +  ' passed date');
          object.set("statutCommerce", 0);
          await object.save(null, {useMasterKey:true});
        }
      }
    }
  });

}

cloud/main.js
// TODO: After save on video to generate video thumbnail & description as 'video presentation {{ nom_du_commerce }}'
// TODO: Devide cloud code into multiple controller class, like Commerce.js (before, after and webhooks)

var NodeGeocoder = require('node-geocoder');
var geocoder = NodeGeocoder({
    provider: 'openstreetmap',
    httpAdapter: 'https',
    formatter: null
});

// Clear associated photos & videos of a commerce when deleted
Parse.Cloud.afterDelete("Commerce", (request) => {
    const commerce = request.object;

    // 1 - Get all photos => delete them all
    const Commerce_Photos = Parse.Object.extend("Commerce_Photos");
    const queryPhotos = new Parse.Query(Commerce_Photos);
    queryPhotos.equalTo("commerce", commerce);
    queryPhotos.find().then(photos => {
        Parse.Object.destroyAll(photos).then(() => {
            console.log("Photos deleted with success");
        })
        .catch(error => {
            if (error.code === Parse.Error.AGGREGATE_ERROR) {
                for (var i = 0; i < error.errors.length; i++) {
                    console.log("Couldn't delete " + error.errors[i].object.id +
                    "due to " + error.errors[i].message);
                }
            } else {
                console.log("Delete aborted because of " + error.message);
            }
        });
    })
    .catch((error) => {
        console.error("Error finding related photos " + error.code + ": " + error.message);
    });
    // 2 - Get all videos => delete them all
    const Commerce_Videos = Parse.Object.extend("Commerce_Videos");
    const queryVideos = new Parse.Query(Commerce_Videos);
    queryPhotos.equalTo("leCommerce", commerce);
    queryVideos.find().then(videos => {
        Parse.Object.destroyAll(videos).then(() => {
            console.log("Videos deleted with success");
        })
        .catch(error => {
            if (error.code === Parse.Error.AGGREGATE_ERROR) {
                for (var i = 0; i < error.errors.length; i++) {
                    console.log("Couldn't delete " + error.errors[i].object.id +
                    "due to " + error.errors[i].message);
                }
            } else {
                console.log("Delete aborted because of " + error.message);
            }
        });
    })
    .catch((error) => {
        console.error("Error finding related videos " + error.code + ": " + error.message);
    });
});

Parse.Cloud.define('nullPosition', async (request) => {
    const query = new Parse.Query("Commerce");
    query.equalTo("position", new Parse.GeoPoint({latitude: 0.0, longitude: 0.0}) );             // Status of commerce == paid
    const result = await query.find();
    return result;
});

Parse.Cloud.afterSave(Parse.User, (request) => {
    const parseObject   = request.object;
    const partagesIds   = parseObject.get("mes_partages");

    // Set mespartages to empty arrays if null
    if (partagesIds === undefined || partagesIds === null || typeof partagesIds === 'undefined')  {
        parseObject.set("mes_partages", []);
        parseObject.set("mes_partages_dates", []);
    }
});

Parse.Cloud.afterSave("Commerce", (request) => {
    const parseObject   = request.object;
    const description   = parseObject.get("description");
    const brouillon     = parseObject.get("brouillon");
    const position      = parseObject.get("position");
    const endOfPayment  = parseObject.get("endedSubscription");
    const sessions      = parseObject.get("stripeCheckoutSession");

    if (typeof position === 'undefined' || position === undefined || position === null || (position.latitude === 0 && position.longitude === 0)) {
        const address = parseObject.get("adresse");
        geocoder.geocode(address)
                .then(response => {
                    const responseArray = JSON.parse(JSON.stringify(response));
                    if (responseArray.length !== 0) {
                        const firstResponse = responseArray[0];
                        
                        parseObject.set("position", new Parse.GeoPoint({
                            latitude: firstResponse.latitude, 
                            longitude: firstResponse.longitude
                        }));
                    }
                })
                .catch(error => { console.log(error); });
    }

    // Set blank and empty values
    if (brouillon === undefined || brouillon === null || typeof brouillon === 'undefined')  {
        parseObject.set("brouillon", true);
    }
    if (endOfPayment === undefined || endOfPayment === null || typeof endOfPayment === 'undefined')  {
        parseObject.set("endedSubscription", new Date());
    }
    if (sessions === undefined || sessions === null || typeof sessions === 'undefined') {
        parseObject.set("stripeCheckoutSession", []);
    }
});

Parse.Cloud.define('endedSubscription', async (request) => {
    const query = new Parse.Query("Commerce");

    query.exists("endSubscription");                // End of subscription is not null
    query.lessThan("endSubscription", new Date());  // Current date is after subscription date 
    query.equalTo("statutCommerce", 1);             // Status of commerce == paid

    const result = await query.find();
    return result;
});

// Parse.Cloud.define("sendOutdatedEmail", (request) => {
//   // Get access to Parse Server's cache
//   const { AppCache } = require('parse-server/lib/cache');
//     // Get a reference to the MailgunAdapter
//     // NOTE: It's best to do this inside the Parse.Cloud.define(...) method body and not at the top of your file with your other imports. This gives Parse Server time to boot, setup cloud code and the email adapter.
//     const MailgunAdapter = AppCache.get('JVQZMCuNYvnecPWvWFDTZa8A').userController.adapter;

//     // request.params.update_at

//     // Invoke the send method with an options object
//     MailgunAdapter.send({
//         templateName: 'customEmailAlert',
//       // Optional override of your configuration's subject
//       subject: 'Important: action required',
//       // Optional override of the adapter's fromAddress
//       fromAddress: 'Alerts <noreply@yourapp.com>',
//       recipient: 'user@email.com',
//       variables: { alert: 'New posts' }, // {{alert}} will be compiled to 'New posts'
//       // Additional message fields can be included with the "extra" option
//       // See https://nodemailer.com/extras/mailcomposer/#e-mail-message-fields for an overview of what can be included
//       extra: {
//           attachments: [], /* include attachment objects */
//           replyTo: 'reply-to-address'
//       }
//   });
// });

routes/stripe.js
module.exports = function(app){

    const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);

    app.post('/create-checkout-session', async (req, res) => {
        const commerceId = req.body.commerceId;
        const userEmail = req.body.userEmail;

        try {
            const session = await stripe.checkout.sessions.create({
                customer_email: `${userEmail}`,
                payment_method_types: ['card'],
                line_items: [
                    { price: process.env.PRICE, quantity: 1 },
                ],
                mode: 'payment',
                success_url: `${process.env.WEB_URL}/success-checkout/{CHECKOUT_SESSION_ID}/${commerceId}`,
                cancel_url: `${process.env.WEB_URL}/aboutcommerce/${commerceId}`,
                metadata: {
                    'commerce_id': `${commerceId}`,
                    'customer_email': `${userEmail}`
                },
            });

            res.json({ id: session.id });
        } catch (error) {
            res.status(400).json({errorMessage: error});
        }
    });

    app.post('/retrieve-checkout-session-status', async (req, res) => {
        const checkoutId = req.body.checkoutId;

        try {
            const session = await stripe.checkout.sessions.retrieve(`${checkoutId}`);

            if (session.object === 'checkout.session') {
                // Check this for result of payment (enum: paid || unpaid || no_payment_required)
                res.json({ payment_status: session.payment_status });  
            } else {
                res.status(400).json({message: "Invalid id, not returning a session object"});
            }
        } catch (error) {
            res.status(400).json({errorMessage: error});
        }
    });

    // const stripeErrorHandling = () => {
    //     switch (err.type) {
    //       case 'StripeCardError':
    //         // A declined card error
    //         err.message; // => e.g. "Your card's expiration year is invalid."
    //         break;
    //       case 'StripeRateLimitError':
    //         // Too many requests made to the API too quickly
    //         break;
    //       case 'StripeInvalidRequestError':
    //         // Invalid parameters were supplied to Stripe's API
    //         break;
    //       case 'StripeAPIError':
    //         // An error occurred internally with Stripe's API
    //         break;
    //       case 'StripeConnectionError':
    //         // Some kind of error occurred during the HTTPS communication
    //         break;
    //       case 'StripeAuthenticationError':
    //         // You probably used an incorrect API key
    //         break;
    //       default:
    //         // Handle any other types of unexpected errors
    //         break;
    //     }
    // }

}

index.js
// require('dotenv').config({ path: './dev.env' });
require('dotenv').config();

let express         = require('express');
let app             = express();

const stripe 		= require('stripe')(process.env.STRIPE_SECRET_KEY);
let ParseServer     = require('parse-server').ParseServer;
let ParseDashboard  = require('parse-dashboard');
let path            = require('path');
let moment          = require('moment');
let Parse           = require('parse/node');
const resolve       = require('path').resolve;
const bodyParser    = require('body-parser'); // Parse incoming request bodies
let GCSAdapter      = require('@parse/gcs-files-adapter');
let mailgun         = require('mailgun-js')({apiKey: process.env.ADAPTER_API_KEY, domain: process.env.ADAPTER_DOMAIN, host: 'api.eu.mailgun.net'});
const cors          = require('cors');

// Parse lib init
Parse.initialize(process.env.APP_ID, null, process.env.MASTER_KEY);
Parse.masterKey = process.env.MASTER_KEY;
Parse.serverURL = process.env.SERVER_URL;

// Configuration Server
app.use(bodyParser.json({ type: 'application/json' }));
app.use(bodyParser.urlencoded({ extended: true }));

moment().format();

let databaseUri = process.env.DATABASE_URI;
if (!databaseUri) { 
	console.log('DATABASE_URI not specified, falling back to localhost.');
} else {
	console.log('DB URI: ' + databaseUri);
}

let options = { allowInsecureHTTP: true };
let dashboard = new ParseDashboard({
	"apps": [{
		"serverURL":  process.env.SERVER_URL,
		"appId":      process.env.APP_ID,
		"masterKey":  process.env.MASTER_KEY,
		"appName":    process.env.APP_NAME,
		"iconName": "logo.png",
		"supportedPushLocales": ["fr"]
	}],
	"iconsFolder": "icons",
	"users": [{
		"user": process.env.ADMIN_USER,
		"pass": process.env.ADMIN_PASSWORD
	}]
}, options);

let gcsOptions = {
	"projectId": "weeclik-1517332083996",
	"keyFilename": resolve(__dirname, "WeeClik-813d8719632d.json"),
	"bucket": "weeclik-1517332083996.appspot.com",
	"bucketPrefix": 'baas_files/',
	"directAccess": false
}
let gcsAdapter = new GCSAdapter(gcsOptions);
let api = new ParseServer({
	databaseURI:        databaseUri || 'mongodb://localhost:27017/weeclik',
	cloud:              process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js',
	appId:              process.env.APP_ID,
	masterKey:          process.env.MASTER_KEY,
	filesAdapter:       gcsAdapter,
	serverURL:          process.env.SERVER_URL,
	publicServerURL:    process.env.PUBLIC_URL,
	appName:            process.env.APP_NAME,
    maxUploadSize:      process.env.MAX_UPLOAD_SIZE || "1024mb", // 1024 MB = 1 Go
    allowClientClassCreation: true,
      // Enable email verification
      // try to use this (avantage langue) // "@ngti/parse-server-mailgun": "^2.4.18",
      verifyUserEmails: process.env.VERIFY_USER_EMAILS || true,
      emailAdapter: {
      	module: 'parse-server-mailgun',
      	options: {
          // The address that your emails come from
          fromAddress: 'Herrick de l\'équipe Weeclik <contact@weeclik.com>',
          // Your domain from mailgun.com
          domain: process.env.ADAPTER_DOMAIN,
          // Mailgun host (default: 'api.mailgun.net'). 
          // When using the EU region, the host should be set to 'api.eu.mailgun.net'
          host: 'api.eu.mailgun.net',
          // Your API key from mailgun.com
          apiKey: process.env.ADAPTER_API_KEY,
          // The template section
          templates: {
          	passwordResetEmail: {
          		subject: 'Mot de passe oublié ?',
          		pathPlainText: resolve(__dirname, 'email_templates/password_reset/password_reset_email.txt'),
          		pathHtml: resolve(__dirname, 'email_templates/password_reset/password_reset_email.html'),
          		callback: (user) => { return { 
          			name: user.get('name') 
          		}}
                  // Now you can use {{name}} in your templates
              },
              verificationEmail: {
              	subject: 'Bienvenue dans la famille Weeclik',
              	pathPlainText: resolve(__dirname, 'email_templates/verification_email/verification_email.txt'),
              	pathHtml: resolve(__dirname, 'email_templates/verification_email/verification_email.html'),
              	callback: (user) => { return { 
              		name: user.get('name') 
              	}}
                  // Now you can use {{name}} in your templates
              },
              customEmailAlert: {
              	subject: 'Message urgent',
              	pathPlainText: resolve(__dirname, 'email_templates/custom_email/custom_email.txt'),
              	pathHtml: resolve(__dirname, 'email_templates/custom_email/custom_email.html'),
              	callback: (user) => { return { 
              		name: user.get('name') 
              	}}
                  // Now you can use {{name}} in your templates
              }
          }
      }
  }
});
// Serve static assets from the /public folder
app.use('/public', express.static(path.join(__dirname, '/public')));
app.get('/apple-app-site-association', (req, res) => {
	res.status(200).sendFile(path.join(__dirname, 'apple-app-site-association'));
});

// Serve the Parse API on the /parse URL prefix
let mountPath = process.env.PARSE_MOUNT;
app.use(mountPath, api);

// make the Parse Dashboard available at /dashboard
app.use('/dashboard', dashboard);

// Allow all cors origin
app.use(cors());

let port = process.env.PORT || 1337;
let httpServer = require('http').createServer(app);
httpServer.listen(port, () => { console.log('Running weeclik server on port ' + port); });

// Import routes
require('./routes/stripe')(app);
// Cron tasks (repetitive tasks)
require('./cron/cron')(app);

app.post('/send-error-mail', (req, res) => {
	console.log(req.body);
	if (req.body.content_message) {
		const data = {
			from: 'iOS APP <no-reply@weeclik.com>',
			to: 'contact@herrick-wolber.fr',
			subject: 'Error in iOS App',
			text: req.body.content_message
		};

		mailgun.messages().send(data, (error, body) => {
			if (error) {
				return error
			} else {
				return res.status(200).send({
					success: 'true',
					message: 'Message envoyé avec succès',
				});
			}
			console.log(body);

		});
	} else {
	    return res.status(422).send({
	    	success: 'false',
	    	message: 'Missing parameter : content_message is undefined'
	    })
	}
});

// Match the raw body to content type application/json
// app.post('/webhook', (request, response) => {
// 	let event;
// 	try {event = request.body;}
// 	catch (err) {response.status(400).send(`Webhook Error: ${err.message}`);}
// 	console.log(event)
//     // Handle the event
//     switch (event.type) {
//     	case 'payment_intent.succeeded':
//     	const paymentIntent = event.data.object;
//     	response.json({received: true, method : event.data.object});
//     	break;
//     	case 'payment_method.attached':
//     	response.json({received: true, method : event.data.object});
//     	break;
//     	case 'payment_intent.created':
//     	response.json({received: true, method : event.data.object});
//     	break;
//     	case 'charge.succeeded':
//     	response.json({received: true, method : event.data.object});
//     	break;
//     	default:
//         // Unexpected event type
//         console.log('Stripe event not handled by server')
//         console.log(event.type)
//         return response.status(400).end();
//     }
// });

app.post("/publish-commerce", async (req, res) => {
	console.log("/publish-commerce commerce endpoint");

	const commerceId = req.body.commerceId;
	const checkoutSessionId = req.body.checkoutSessionId;
	const userId = req.body.userId;
	console.log("User "+ userId +" called checkoutSessionId: "+ checkoutSessionId + " for commerce: " + commerceId);
	
	if (commerceId && checkoutSessionId && userId) {
		var query = new Parse.Query(Parse.Object.extend("StripeCheckoutSessions"));
		query.equalTo("sessionId", checkoutSessionId);
		const result = await query.find();
		if (result === undefined || result.length === 0) {
			var queryR = new Parse.Query(Parse.Object.extend("Commerce"));
			queryR.get(commerceId).then( async (commerce) => {
				var sessions = commerce.get("stripeCheckoutSession");
				// Set empty value if null/undefined
				if (sessions === undefined) { commerce.set("stripeCheckoutSession", []); }
				if (sessions === undefined || !sessions.includes(checkoutSessionId)) {
					// Everything is ok continue with commerce publish
					try {
						const session = await stripe.checkout.sessions.retrieve(`${checkoutSessionId}`);
						if (session.object === 'checkout.session') {
							// Check this for result of payment (enum: paid || unpaid || no_payment_required)
							if (session.payment_status === 'paid') {
								// Save checkout session id to prevent fraud
								const StripeSessions = Parse.Object.extend("StripeCheckoutSessions");
								const sessionObject = new StripeSessions();
								sessionObject.set("sessionId", checkoutSessionId);
								sessionObject.set("commerceId", commerceId);
								sessionObject.set("commerce", commerce);
								const acl = new Parse.ACL();
                        		acl.setPublicReadAccess(true);
								acl.setRoleWriteAccess("admin", true);
								acl.setWriteAccess(userId, true);
								sessionObject.setACL(acl);
								sessionObject.save(null, {useMasterKey: true}).then((_) =>{
									// Update commerce data
									var aYearFromNow = new Date();
									aYearFromNow.setFullYear(aYearFromNow.getFullYear() + 1);
									commerce.set("endedSubscription", aYearFromNow);
									commerce.set("statutCommerce", 1);
									commerce.set("brouillon", false);
									commerce.addUnique("stripeCheckoutSession", checkoutSessionId);
									commerce.save(null, {useMasterKey: true}).then((_) => {
										return res.status(200).send({ message: 'Publishing of commerce: '+ commerceId + 'has been done successfully' });
									}, (savingError) => {
										return res.status(402).send({ error: 'Updating of commerce did fail. Original error => '+ savingError.message });
									});
								}, (sessionError) => {
									return res.status(403).send({ error: 'Error while saving the session object. Original error => '+ sessionError });
								});
							} else {
								res.status(403).json({ error: `Checkout session status: ${session.payment_status}, publishing not allowed for commerce: ${commerceId}`});
							}
						} else {
							res.status(402).json({ error: "Invalid chackout session id provided, not returning a session object"});
						}
					} catch (error) {
						res.status(400).json({ error: error});
					}
				} else {
					// Stripe Checkout Session already exists in database
					return res.status(403).send({ error: 'Stripe Checkout Session already exists in database'});
				}
			}, (commerceError) => {
				return res.status(403).send({ error: 'Commerce not found. Original error => '+ commerceError.message });
			});

		} else {
			return res.status(403).send({ error: 'Stripe Checkout Session Id already exists in database'});
		}

	} else {
		res.status(400).json({ error: "The request was unacceptable, due to missing a required parameter."});
	}
});

app.post("/share", (req, res) => {    
	console.log("/share commerce endpoint");
	console.log("User: "+ req.body.userId + " is sharing commerce: " + req.body.commerceId);
	const commerceId = req.body.commerceId;
	const userId = req.body.userId;

	if (commerceId) {
		var query = new Parse.Query(Parse.Object.extend("Commerce"));
		query.get(commerceId).then((commerce) => {
			commerce.increment("nombrePartages");
			commerce.save(null, {useMasterKey: true}).then((commerceSaved) => {
				// Fetch user
				if (userId) {
					var userQuery = new Parse.Query(Parse.User);
					userQuery.get(userId).then((user) => {
						// TODO: check if values are nil
						var shares = user.get("mes_partages");
						if (shares === undefined || shares === null || typeof shares === 'undefined') {
							// Date is null
							user.set("mes_partages", []);
							user.set("mes_partages_dates", []);
						} else {
							// Data already exists
							user.add("mes_partages", commerce.id);
							user.add("mes_partages_dates", new Date());
						}

						user.save(null, {useMasterKey: true}).then((user) => {
							return res.status(200).send({ message: 'Sharing of commerce: '+ commerceId + 'has been done successfully by user: '+ userId });
						}, (errorSavingUser) => {
							return res.status(402).send({ error: 'Updating of user did fail. Original error => '+ errorSavingUser.message });
						});
					}, (userError) => {
						return res.status(404).send({ error: 'User not found. Original error => '+ userError.message });
					});
				} else {
					return res.status(200).send({ message: 'Sharing of commerce: '+ commerceId + 'has been done successfully' });
				}
				
			}, (savingError) => {
				return res.status(401).send({ error: 'Updating of commerce did fail. Original error => '+ savingError.message });
			});
		}, (commerceError) => {
			return res.status(403).send({ error: 'Commerce not found. Original error => '+ commerceError.message });
		});
	} else {
		return res.status(400).send({ error: 'missing commerce id parameter' });
	}

});